File: backend/trpc/routers/projects.ts
/**
 * ================================================
 * tRPC ROUTER — Projects
 * ================================================
 */

import { z } from 'zod';
import { router, publicProcedure } from '../init';
import testmoService from '../../services/testmo.service';

const projectIdInput = z.object({
  projectId: z.number().int().positive(),
});

const runsInput = z.object({
  projectId: z.number().int().positive(),
  active: z.boolean().optional().default(true),
});

export const projectsRouter = router({
  list: publicProcedure.query(async () => {
    const projectsRaw = await testmoService.getProjects();
    const projects = Array.isArray(projectsRaw) ? projectsRaw : (projectsRaw as any)?.result || [];
    return { success: true as const, data: { result: projects }, timestamp: new Date().toISOString() };
  }),

  runs: publicProcedure
    .input(runsInput)
    .query(async ({ input }) => {
      const runs = await testmoService.getProjectRuns(input.projectId, input.active);
      return { success: true as const, data: runs, timestamp: new Date().toISOString() };
    }),

  milestones: publicProcedure
    .input(projectIdInput)
    .query(async ({ input }) => {
      const milestones = await testmoService.getProjectMilestones(input.projectId);
      return { success: true as const, data: milestones, timestamp: new Date().toISOString() };
    }),

  automation: publicProcedure
    .input(projectIdInput)
    .query(async ({ input }) => {
      const automationRuns = await testmoService.getAutomationRuns(input.projectId);
      return { success: true as const, data: automationRuns, timestamp: new Date().toISOString() };
    }),
});

File: backend_py/app/routers/projects.py
"""Projects, runs, milestones, automation."""

from __future__ import annotations

from fastapi import APIRouter, Query

from app.deps import DBMain
from app.services.testmo import testmo_service

router = APIRouter()


@router.get("/")
async def list_projects(db: DBMain):
    projects = await testmo_service.get_projects()
    return {"projects": projects}


@router.get("/{project_id}/runs")
async def get_project_runs(project_id: int, active: bool = Query(False), db: DBMain = None):
    runs = await testmo_service.get_project_runs(project_id, active_only=active)
    return {"runs": runs}


@router.get("/{project_id}/milestones")
async def get_project_milestones(project_id: int, db: DBMain):
    milestones = await testmo_service.get_project_milestones(project_id)
    return {"success": True, "data": {"result": milestones}}


@router.get("/{project_id}/automation")
async def get_project_automation(project_id: int, db: DBMain):
    automation = await testmo_service.get_automation_runs(project_id)
    return {"automation": automation}
